### ISSUE TITLE
The optimizer's best learning rate and momentum 

### ISSUE THREAD

**Issue by bekhzod-olimov** [ultralytics/ultralytics#17460]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hey there! Thank you very much for your excellent work in addressing all the issues in this repository.  I have a question regarding the optimizer settings. By default, it automatically selects the optimizer and it's parameters. 

`optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically...`

I am wondering how the best learning rate and moment values are chosen? Could you please provide the chunk of code where the best values are selected? 

Thank you very much for your reply in advance.

Best Regards,
Bekhzod

**Comment by UltralyticsAssistant** [comment]
👋 Hello @bekhzod-olimov, thank you for your question and for using Ultralytics 🚀! 

We highly recommend checking out our comprehensive [Docs](https://docs.ultralytics.com) which cover many essential aspects, including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples. These resources can be instrumental in understanding optimizer settings and much more.

If your inquiry pertains to a 🐛 Bug Report, please assist us by providing a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to streamline the debugging process.

For custom training settings and ❓ Questions, supplying detailed information is crucial. This includes dataset samples, training logs, and ensuring adherence to our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Feel free to engage with the Ultralytics community! Whether you prefer the dynamic environment of [Discord](https://ultralytics.com/discord) 🎧, detailed discussions on [Discourse](https://community.ultralytics.com), or vibrant exchanges on our [Subreddit](https://reddit.com/r/ultralytics), there’s a space for you!

### Upgrade

To ensure you’re on the leading edge with the latest `ultralytics` package, please upgrade in an environment with all necessary [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml):

```bash
pip install -U ultralytics
```

### Environments

Utilize YOLO across various up-to-date environments:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

### Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

Our CI tests are conducted daily and following every commit to assure optimal YOLO functionality across different platforms. A green badge signifies all tests are currently passing.

🔄 This is an automated response, but rest assured, an Ultralytics engineer will review your question and provide additional assistance soon. Thank you for your patience and understanding!

**Comment by Y-T-G** [comment]
https://github.com/ultralytics/ultralytics/blob/ed252226a9d8d3009ead5a511ab6ba51dafa132a/ultralytics/engine/trainer.py#L753

**Comment by HimanshuJanbandhu** [comment]
So once lr_fit is calculated, exactly what learning rate scheduler is used, as by default cosine_lr is `False`

**Comment by glenn-jocher** [comment]
The default learning rate scheduler used in Ultralytics YOLO is a linear warmup followed by a constant learning rate, unless specified otherwise by setting `cos_lr=True` for a cosine learning rate schedule.

**Comment by furycurry** [comment]
@glenn-jocher @Y-T-G won't it be possible to set a desired lr0?

**Comment by Y-T-G** [comment]
You can set it if you also specify an optimizer

**Comment by furycurry** [comment]
> You can set it if you also specify an optimizer

thanks

**Comment by glenn-jocher** [comment]
You're welcome! Let us know if you have any further questions.
